src/App.tsx

```diff
@@ -1,26 +1,235 @@
+
+    return () => clearTimeout(timer);
+  }, [errorMessage]);
+
+  useEffect(() => {
+    focusInputField();
+  }, [todoTitle, todos, selectedTodoStatus, isLoading]);
+
+  const filteringTodosByActiveStatus = useMemo(
+    () => [...todos].filter(todo => !todo.completed),
+    [todos],
+  );
+
+  const filteringTodosByCompletedStatus = useMemo(
+    () => [...todos].filter(todo => todo.completed),
+    [todos],
+  );
+
+  const filteredTodos = useMemo(() => {
+    switch (selectedTodoStatus) {
+      case TodoStatus.Active:
+        return filteringTodosByActiveStatus;
+
+      case TodoStatus.Completed:
+        return filteringTodosByCompletedStatus;
+
+      default:
+        return todos;
+    }
+  }, [
+    filteringTodosByActiveStatus,
+    filteringTodosByCompletedStatus,
+    selectedTodoStatus,
+    todos,
+  ]);
+
+  const changeTodoTitleHandler = useCallback(
+    (e: ChangeEvent<HTMLInputElement>) => {
+      setErrorMessage('');
+      setTodoTitle(e.target.value);
+    },
+    [],
+  );
+
+  const closeErrorHandler = () => {
+    setErrorMessage('');
+  };
+
+  const handleStatusChange = (status: TodoStatus) => {
+    setSelectedTodoStatus(status);
+  };
+
+  const addTodo = useCallback(
+    async (e: React.FormEvent<HTMLFormElement>) => {
+      e.preventDefault();
+
+      if (!todoTitle.trim().length) {
+        setErrorMessage(ErrorType.EMPTY_TITLE);
+
+        return;
+      }
+
+      setIsLoading(true);
+      setErrorMessage('');
+
+      const newTempTodo: Todo = { ...emptyTodo, title: todoTitle.trim() };
+
+      setTempTodo(newTempTodo);
+      setProcessingTodos([newTempTodo.id]);
+
+      try {
+        const todo = await uploadTodo({
+          ...emptyTodo,
+          title: todoTitle.trim(),
+        });
+
+        setTodos(currentTodos => [...currentTodos, todo]);
+        setTodoTitle('');
+      } catch {
+        setErrorMessage(ErrorType.ADD_TODO);
+      } finally {
+        setIsLoading(false);
+        setTempTodo(null);
+        setProcessingTodos([]);
+        focusInputField();
+      }
+    },
+    [todoTitle],
+  );
+
+  const removeTodo = useCallback((id: number) => {
+    setErrorMessage('');
+    setProcessingTodos((prev: number[]) => [...prev, id]);
+
+    deleteTodo(id)
+      .then(() =>
+        setTodos(currentTodos => currentTodos.filter(todo => todo.id !== id)),
+      )
+      .catch(() => setErrorMessage(ErrorType.DELETE_TODO))
+      .finally(() => {
+        setProcessingTodos((prev: number[]) =>
+          prev.filter(prevItem => prevItem !== id),
+        );
+        focusInputField();
+      });
+  }, []);
+
+  const removeTodos = useCallback(async () => {
+    setIsLoading(true);
+    setErrorMessage('');
+
+    const deletePromises = filteringTodosByCompletedStatus.map(todo => {
+      setProcessingTodos((prev: number[]) => [...prev, todo.id]);
+
+      deleteTodo(todo.id)
+        .then(() => {
+          setTodos(currentTodos => currentTodos.filter(t => t.id !== todo.id));
```

Comment:
Its a good practice to use more descriptive names
```suggestion
          setTodos(currentTodos => currentTodos.filter(currentTodo => currentTodo.id !== todo.id));
```